rfcs/71-write-only-world-query.md

```diff
@@ -0,0 +1,385 @@
+# `WriteOnly` world query
+
+## Summary
+
+Add a `WriteOnly<T>` `WorldQuery`. `WriteOnly` acts like `&mut`, but tells bevy
+that the value of component `T` is overwritten unconditionally by the system.
+Add the `B0006` error. This tells users that they are modifying a component
+that is later overwritten unconditionally.
+
+## Motivation
+
+When I write to the `Transform` of a UI node, it doesn't do anything
+and I see no error messages.
+
+Why this happens is fairly trivial: `Transform` is updated by the UI system
+in `PostUpdate`, and overwrites the user-set value unconditionally.
+
+This is also true of `AnimationClip`s. A system overwrites bone transforms
+targeted by an animations in `PostUpdate`. Even if the `AnimationClip` is
+paused!
+
+## User-facing explanation
+
+### A new `WorldQuery`
+
+We introduce a new `WorldQuery`: `WriteOnly<T>`.
+
+```rust
+struct WriteOnly<'a, T>(Mut<'a, T>);
+
+impl<'a, T> WriteOnly<'a, T> {
+  pub fn set(&mut self, value: T) {
+    *self.0 = value
+  }
+  /// Only **write** to the returned reference.
+  /// If you have a `Vec`, for example, you should not leave any of the
+  /// previously existing value.
+  pub fn ref_mut(&mut self) -> Mut<T> {
+    self.0.reborrow()
+  }
+  pub fn into_inner(self) -> Mut<'a, T> {
+    self.0
+  }
+}
+
+```
+
+`WriteOnly<T>` is very much nothing more than a thin wrapper around `Mut`.
```

Comment:
`WriteOnly` seems like the wrong name for this:
* We already have the concept of "read only queries". Which means "you statically do not have permission to do anything but read". This is not the "write" equivalent of that.
* The goal of this feature _is not_ to assert that we only write to a thing (and not read from it). It it is to assert that _all_ values will be written to, with none skipped. Reading before or after the write is completely ok, provided you do the write. `WriteOnly` not only isn't accurate ... it encourages the user to think about it in the wrong way.  If I saw WriteOnly in a query, I wouldn't think "i need to write every item in the query". I would think "i cannot read items in this query but I can still skip items I don't want to write".

Reply:
The no-read semantic is important. What makes `ui_layout_system` handling of `Transform` different from other systems is indeed that it overwrites the `Transform` without leaving a trace of the existing value. Whether it does this to all the entities it access or only a subset is only partially relevant.

```rust
fn sys(mut q: Query<&mut Transform>) {
 for t in &mut q {
    t.translation.x += 1.0;
  }
}
```

Here, a previous system that changes `Transform` still has a visible effect after `sys` runs. So we don't want to say that translation was overwritten.

While for the following:

```rust
fn sys(mut q: Query<&mut Transform>) {
 for t in &mut q {
    if t.translation.x > 10.0 {
      continue;
    }
    *t = Transform::IDENTITY;
  }
}
```

The previous system's value will be overwritten (for a subset of `Transform`) If the user wrote a system that sets a `Transform` that has a `translation.x` smaller than 10.0, that value will never be read afterward.

Our goal here is an error message similar to rust's [`unused-assignments`](https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unused-assignments) lint.